### DISCUSSION TITLE
How to shut the server down

### DISCUSSION THREAD

**Discussion by kwalcock** [lihaoyi/cask#132] [Q&A] [answered]
From within my program, how can I shut down the server that is running?  There must be a thread that is still alive when my program is otherwise finished and I need to terminate it so that my program can end.

**Comment by jodersky** [comment, marked as answer]
TL;DR: if you use cask with `object App extend cask.MainRoutes {...}` you can't do it

**In general**, you need to access the Undertow server that underpins cask's HTTP handling and call the `stop()` method.

As per [the Undertow docs](https://undertow.io/undertow-docs/undertow-docs-2.1.0/index.html#the-builder-api)

> The above example starts a simple server that returns Hello World to all requests. The server will listen on the localhost address on port 8080 until the server.stop() method is called.

Now, how you access the server is dependent on how you use cask.

**In case you started cask by extending `cask.MainRoutes` and calling `main()`** (as is done in the most simple cases e.g. `object App extends cask.MainRoutes`), then there is no way to stop the server programmatically. This is because cask's [`main()` method](https://github.com/com-lihaoyi/cask/blob/master/cask/src/cask/main/Main.scala#L63-L70) doesn't return the server instance, so you can't call `stop()` on it.

**What you could do instead** is reimplement what the main method does in a more fine grained way, specific to your app, and call that instead. For example, here is an extract of how I commonly use cask when I need more sophisticated setup and teardown code:

```scala
class HttpServer(
  override val host: String,
  override val port: Int,
  override val debugMode: Boolean,
  val routes: Iterable[cask.Routes]
) extends cask.MainRoutes:
  override def allRoutes = routes.toSeq

  val server =
    io.undertow.Undertow.builder
      .addHttpListener(port, host)
      .setHandler(defaultHandler)
      .build()

  def start() = server.start()

  def stop() = server.stop()

end HttpServer

def main() =
  val server = HttpServer(....)

  println("starting http server")
  server.start()

  // wait until exit condition
  // ...

  println("stopping http server")
  server.stop()
```

**Comment by kwalcock** [comment]
Thank you!  Instead of calling main from my extension to MainRoutes, I now just make my own server and ask it to start.  Then I stop it when it's time to close.  It's easy enough to implement once one knows that to do.
